File: server/routes/sessions.py
"""Persistent Session browsing and metadata routes."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel, Field, field_validator

from server.errors import APIError
from server.services.agent_registry import AgentRegistry
from server.services.session_merge import append_merge_summary, create_session_merge_summary
from server.services.session_store import (
    SessionStore,
    session_detail,
    session_info_list_to_dict,
)

router = APIRouter(prefix="/api/sessions", tags=["sessions"])


class RenameSessionRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=200)]

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Session name must not be blank")
        return normalized


class MergeSessionRequest(BaseModel):
    sourceSessionId: str


class ForkSessionRequest(BaseModel):
    leafId: Annotated[str, Field(min_length=1)]


def get_session_store(request: Request) -> SessionStore:
    return request.app.state.session_store


def get_agent_registry(request: Request) -> AgentRegistry:
    return request.app.state.agent_registry


@router.get("")
async def list_sessions(store: SessionStore = Depends(get_session_store)) -> dict:
    return {"sessions": session_info_list_to_dict(store.list())}


@router.get("/{session_id}")
async def get_session(session_id: str, store: SessionStore = Depends(get_session_store)) -> dict:
    info = store.find(session_id)
    if info is None:
        raise APIError(404, "session_not_found", f"Session {session_id!r} was not found")
    manager = store.open(session_id)
    if manager is None:  # The file may have disappeared between discovery and opening.
        raise APIError(404, "session_not_found", f"Session {session_id!r} was not found")
    serialized = session_info_list_to_dict(store.list())
    parent_session_id = next(
        (item["parentSessionId"] for item in serialized if item["id"] == session_id),
        None,
    )
    return session_detail(manager, info, parent_session_id=parent_session_id)


@router.get("/{session_id}/context")
async def get_session_context(
    session_id: str,
    leaf_id: str | None = None,
    store: SessionStore = Depends(get_session_store),
) -> dict:
    manager = store.open(session_id)
    if manager is None:
        raise APIError(404, "session_not_found", f"Session {session_id!r} was not found")
    if leaf_id is not None:
        try:
            manager.branch(leaf_id)
        except KeyError as exception:
            raise APIError(
                404,
                "entry_not_found",
                f"Entry {leaf_id!r} was not found in Session {session_id!r}",
            ) from exception
    return manager.build_web_session_context()


@router.patch("/{session_id}")
async def rename_session(
    session_id: str,
    body: RenameSessionRequest,
    store: SessionStore = Depends(get_session_store),
) -> dict[str, bool]:
    manager = store.open(session_id)
    if manager is None:
        raise APIError(404, "session_not_found", f"Session {session_id!r} was not found")
    manager.append_session_info(body.name)
    return {"ok": True}


@router.delete("/{session_id}")
async def delete_session(
    session_id: str,
    store: SessionStore = Depends(get_session_store),
    registry: AgentRegistry = Depends(get_agent_registry),
) -> dict:
    if store.find(session_id) is None:
        raise APIError(404, "session_not_found", f"Session {session_id!r} was not found")
    await registry.remove(session_id)
    try:
        reparented = store.delete_with_reparent(session_id)
    except OSError as exception:
        raise APIError(500, "session_delete_failed", str(exception)) from exception
    if reparented is None:
        raise APIError(404, "session_not_found", f"Session {session_id!r} was not found")
    return {"ok": True, "reparentedCount": reparented}


@router.post("/{session_id}/fork")
async def fork_session(
    session_id: str,
    body: ForkSessionRequest,
    store: SessionStore = Depends(get_session_store),
) -> dict:
    source = store.open(session_id)
    if source is None:
        raise APIError(404, "session_not_found", f"Session {session_id!r} was not found")
    try:
        new_path = source.create_branched_session(body.leafId)
    except KeyError as exception:
        raise APIError(
            404,
            "entry_not_found",
            f"Entry {body.leafId!r} was not found in Session {session_id!r}",
        ) from exception
    if new_path is None or not new_path.exists():
        raise APIError(
            409,
            "fork_not_persisted",
            "The selected branch has no assistant response and cannot be persisted",
        )
    info = store.find(source.session_id)
    if info is None:
        raise APIError(500, "session_fork_failed", "The forked Session could not be discovered")
    serialized = session_info_list_to_dict(store.list())
    forked_info = next(item for item in serialized if item["id"] == source.session_id)
    return {"ok": True, "sessionId": source.session_id, "info": forked_info}


@router.post("/{session_id}/merge")
async def merge_session(
    session_id: str,
    body: MergeSessionRequest,
    store: SessionStore = Depends(get_session_store),
    registry: AgentRegistry = Depends(get_agent_registry),
) -> dict:
    if body.sourceSessionId == session_id:
        raise APIError(400, "invalid_merge", "A Session cannot be merged into itself")
    target_info = store.find(session_id)
    source = store.open(body.sourceSessionId)
    if target_info is None:
        raise APIError(404, "session_not_found", f"Target Session {session_id!r} was not found")
    if source is None:
        raise APIError(
            404,
            "session_not_found",
            f"Source Session {body.sourceSessionId!r} was not found",
        )
    live_target = registry.get(session_id)
    target = live_target.agent.session_manager if live_target is not None else store.open(session_id)
    if target is None:
        raise APIError(404, "session_not_found", f"Target Session {session_id!r} was not found")
    summary = create_session_merge_summary(source, target)
    if summary is None:
        raise APIError(409, "nothing_to_merge", "Source Session has no new mergeable content")
    entry_id = append_merge_summary(target, body.sourceSessionId, summary)
    if live_target is not None:
        live_target.agent.messages = target.build_session_context()["messages"]
    return {
        "ok": True,
        "entryId": entry_id,
        "sourceUniqueEntryCount": summary.source_unique_entry_count,
        "summarizedItemCount": summary.summarized_item_count,
    }

File: node-server/src/routes/sessions.ts
/** Persistent original-Pi session discovery and detail routes. */

import type { FastifyPluginAsync } from "fastify";
import { SessionManager } from "@earendil-works/pi-coding-agent";
import { readFile, rename, rm, writeFile } from "node:fs/promises";

import { AgentRegistry, type PersistedSessionInfo, type RegistryEntry } from "../services/agent-registry.js";
import { ApiError } from "../errors.js";
import { appendMergeSummary, createMergeSummary, type MergeableSessionManager } from "../services/session-merge.js";

export interface SessionRouteOptions {
  registry: AgentRegistry;
}

function pathKey(path: string | undefined): string | undefined {
  return path?.replaceAll("\\", "/").toLocaleLowerCase();
}

function serializeInfo(
  session: PersistedSessionInfo,
  parentIds: Map<string, string>,
): Record<string, unknown> {
  return {
    id: session.id,
    path: session.path || null,
    cwd: session.cwd,
    name: session.name ?? null,
    created: session.created.toISOString(),
    modified: session.modified.toISOString(),
    messageCount: session.messageCount,
    firstMessage: session.firstMessage,
    parentSessionId: pathKey(session.parentSessionPath) === undefined
      ? null
      : parentIds.get(pathKey(session.parentSessionPath)!) ?? null,
    parentSessionPath: session.parentSessionPath ?? null,
  };
}

async function sessionIndex(registry: AgentRegistry): Promise<{
  sessions: Array<PersistedSessionInfo & { active: boolean }>;
  parentIds: Map<string, string>;
}> {
  const sessions = await registry.listSessions();
  const parentIds = new Map<string, string>();
  for (const session of sessions) {
    const key = pathKey(session.path);
    if (key) parentIds.set(key, session.id);
  }
  return { sessions, parentIds };
}

function fallbackInfo(entry: RegistryEntry): PersistedSessionInfo {
  return {
    id: entry.session.sessionId,
    path: entry.session.sessionManager?.getSessionFile() ?? "",
    cwd: entry.cwd,
    created: entry.createdAt,
    modified: entry.createdAt,
    messageCount: entry.session.messages.length,
    firstMessage: "",
  };
}

async function reparentAndDelete(target: PersistedSessionInfo, sessions: PersistedSessionInfo[]): Promise<number> {
  const targetKey = pathKey(target.path);
  const children = sessions.filter((session) => pathKey(session.parentSessionPath) === targetKey);
  for (const child of children) {
    const content = await readFile(child.path, "utf8");
    const lines = content.split(/\r?\n/);
    if (!lines[0]) throw new ApiError(409, "session_delete_failed", `Session header is missing: ${child.id}`);
    let header: Record<string, unknown>;
    try { header = JSON.parse(lines[0]) as Record<string, unknown>; }
    catch { throw new ApiError(409, "session_delete_failed", `Session header is invalid: ${child.id}`); }
    if (target.parentSessionPath) header.parentSession = target.parentSessionPath;
    else delete header.parentSession;
    lines[0] = JSON.stringify(header);
    const temporary = `${child.path}.${crypto.randomUUID()}.tmp`;
    await writeFile(temporary, lines.join("\n"), "utf8");
    await rename(temporary, child.path);
  }
  await rm(target.path);
  return children.length;
}

export const sessionRoutes: FastifyPluginAsync<SessionRouteOptions> = async (app, options) => {
  app.get("/", async () => {
    const { sessions, parentIds } = await sessionIndex(options.registry);
    return { sessions: sessions.map((session) => serializeInfo(session, parentIds)) };
  });

  app.post<{ Params: { sessionId: string } }>("/:sessionId/fork", async (request) => {
    const body = request.body;
    if (body === null || typeof body !== "object" || Array.isArray(body)
      || typeof (body as { leafId?: unknown }).leafId !== "string"
      || !(body as { leafId: string }).leafId.trim()) {
      throw new ApiError(422, "validation_error", "leafId is required");
    }
    const source = await options.registry.open(request.params.sessionId);
    const manager = source.session.sessionManager;
    if (manager === undefined) {
      throw new ApiError(409, "fork_unavailable", "Session branching is unavailable for this session");
    }
    const filePath = (manager as typeof manager & { createBranchedSession(leafId: string): string | undefined })
      .createBranchedSession((body as { leafId: string }).leafId);
    if (filePath === undefined) {
      throw new ApiError(409, "fork_not_persisted", "The selected branch cannot be persisted");
    }
    const forkedId = SessionManager.open(filePath).getSessionId();
    const { sessions, parentIds } = await sessionIndex(options.registry);
    const forked = sessions.find((session) => session.id === forkedId);
    return {
      ok: true,
      sessionId: forkedId,
      info: serializeInfo(forked ?? {
        ...fallbackInfo(source),
        id: forkedId,
        path: filePath,
      }, parentIds),
    };
  });

  app.post<{ Params: { sessionId: string }; Body: { sourceSessionId?: unknown } }>("/:sessionId/merge", async (request) => {
    const sourceSessionId = request.body?.sourceSessionId;
    if (typeof sourceSessionId !== "string" || !sourceSessionId.trim()) {
      throw new ApiError(422, "validation_error", "sourceSessionId is required");
    }
    if (sourceSessionId === request.params.sessionId) {
      throw new ApiError(400, "invalid_merge", "A session cannot be merged into itself");
    }
    const [target, source] = await Promise.all([
      options.registry.open(request.params.sessionId),
      options.registry.open(sourceSessionId),
    ]);
    if (target.session.isStreaming || source.session.isStreaming) {
      throw new ApiError(409, "session_busy", "Stop both sessions before merging");
    }
    const targetManager = target.session.sessionManager as MergeableSessionManager | undefined;
    const sourceManager = source.session.sessionManager as MergeableSessionManager | undefined;
    if (!targetManager || !sourceManager) {
      throw new ApiError(409, "merge_unavailable", "Session merge is unavailable for this session");
    }
    const summary = createMergeSummary(sourceManager, targetManager, sourceSessionId);
    if (!summary) throw new ApiError(409, "nothing_to_merge", "Source session has no mergeable content");
    const entryId = appendMergeSummary(targetManager, sourceSessionId, summary);
    options.registry.syncContext(request.params.sessionId);
    return { ok: true, entryId, sourceUniqueEntryCount: summary.sourceUniqueEntryCount, summarizedItemCount: summary.summarizedItemCount };
  });

  app.patch<{ Params: { sessionId: string }; Body: { name?: unknown } }>("/:sessionId", async (request) => {
    const name = request.body?.name;
    if (typeof name !== "string" || !name.trim() || name.trim().length > 200) {
      throw new ApiError(422, "validation_error", "name must be 1 to 200 characters");
    }
    const entry = await options.registry.open(request.params.sessionId);
    const manager = entry.session.sessionManager as (typeof entry.session.sessionManager & { appendSessionInfo(name: string): string }) | undefined;
    if (!manager) throw new ApiError(409, "rename_unavailable", "Session rename is unavailable for this session");
    manager.appendSessionInfo(name.trim());
    return { ok: true };
  });

  app.delete<{ Params: { sessionId: string } }>("/:sessionId", async (request) => {
    const { sessions } = await sessionIndex(options.registry);
    const target = sessions.find((session) => session.id === request.params.sessionId);
    if (!target || !target.path) throw new ApiError(404, "session_not_found", `Session ${request.params.sessionId} was not found`);
    await options.registry.remove(target.id);
    const reparentedCount = await reparentAndDelete(target, sessions);
    return { ok: true, reparentedCount };
  });

  app.get<{ Params: { sessionId: string } }>("/:sessionId", async (request) => {
    const [entry, index] = await Promise.all([
      options.registry.open(request.params.sessionId),
      sessionIndex(options.registry),
    ]);
    const persisted = index.sessions.find((session) => session.id === entry.session.sessionId)
      ?? entry.persisted
      ?? fallbackInfo(entry);
    const manager = entry.session.sessionManager;
    const context = manager?.buildSessionContext();
    return {
      sessionId: entry.session.sessionId,
      filePath: manager?.getSessionFile() ?? persisted.path ?? null,
      info: serializeInfo(persisted, index.parentIds),
      tree: manager?.getTree() ?? [],
      leafId: manager?.getLeafId() ?? null,
      context: {
        messages: context?.messages ?? entry.session.messages,
        entryIds: [],
        thinkingLevel: context?.thinkingLevel ?? entry.session.thinkingLevel,
        model: context?.model ?? (entry.session.model === undefined
          ? null
          : { provider: entry.session.model.provider, modelId: entry.session.model.id }),
      },
    };
  });
};
